### PR TITLE
Need for (travis) speed

### PR DESCRIPTION
**Pull request by EnTeQuAk**
This is experimental and mostly me playing around with ways to speedup our builds.
Before: around 35-45 minutes
After:
screenshot from 2016-03-17 17-24-32

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,32 +2,70 @@ language: python
...
-  - TOX_ENV=flake8
-  - TOX_ENV=docs
-  - TOX_ENV=assets
-  - TOX_ENV=es
-  - TOX_ENV=addons-devhub-editors
-  - TOX_ENV=main
+  - TOX_ENV=flake8 APP=xx
+  - TOX_ENV=docs APP=xx
+  - TOX_ENV=assets APP=xx
+  - TOX_ENV=elasticsearch APP=xx
```
Comment (EnTeQuAk) [inline .travis.yml]: tox strips empty environment variables so this is a workaround to have `APP` in tox

File: .travis.yml [hunk 2]
```diff
@@ -2,32 +2,70 @@ language: python
...
+  - TOX_ENV=main APP=lib
+  - TOX_ENV=main APP=paypal
+  - TOX_ENV=main APP=reviews
+  - TOX_ENV=main APP=search
+  - TOX_ENV=main APP=signing
+  - TOX_ENV=main APP=stats
+  - TOX_ENV=main APP=tags
+  - TOX_ENV=main APP=translations
+  - TOX_ENV=main APP=users
+  - TOX_ENV=main APP=zadmin
```
Comment (diox) [inline .travis.yml]: I feel like we need a way to automate this, otherwise we **will** forget new apps.

File: .travis.yml [hunk 3]
```diff
@@ -2,32 +2,70 @@ language: python
 sudo: false
 python:
   - "2.7"
+
+cache:
+  directories:
+    - node_modules
+    - $HOME/.cache/pip/
```
Comment (diox) [inline .travis.yml]: Could we even cache the tox directory ? Should we ?

File: setup.cfg [hunk 4]
```diff
@@ -1,5 +1,5 @@
 [pytest]
-addopts = --reuse-db --instafail --showlocals
+addopts =-vvs --reuse-db --instafail --showlocals --tb=short
```
Comment (diox) [inline setup.cfg]: Do we need the `-vvs` ?

**Comment by diox** [comment]
Also, the "before" time is *not* 35/45 minutes in my experience. 

Looking at the most recent master build, it's actually below 15 minutes: https://travis-ci.org/mozilla/addons-server/builds/116681212

File: setup.cfg (hunk 4 above)
Comment (EnTeQuAk) [inline setup.cfg]: I found that the combination -vvs and short tracebacks lead to a pretty good output but I'm open to suggestions.

Comment (diox) [inline setup.cfg]: I would change that in a different PR. I don't really mind, but others might, and more importantly I feel there are other changes we should make to the pytest output (with faster builds, I believe we should revert #1545 and let failures be shown at the bottom - see https://github.com/mozilla/addons-server/issues/1497)

**Comment by EnTeQuAk** [comment]
Alrighty, change of plans: With proper pip and tox caching it looks like we're around 8 minutes total now and only need to split out `addons`, `editors`, `devhub`, `elasticsearch` and the rest should go into one group avoiding the additional setup-time.

File: setup.cfg (hunk 4 above)
Comment (EnTeQuAk) [inline setup.cfg]: ugh, that's what instafail is for! I'll try and see how the results look like without.

**Comment by EnTeQuAk** [comment]
Closing for now, I'll re-open this in a bit with a few more refactorings.

**Comment by diox** [comment]
I just noticed something: we install ES in .travis.yml in `before_install`, but if we only use it in a subset of the tests, we should be able to move the install to the job dealing with that subset, instead of slowing down the install part of each job ?

**Pull request closed without merging**